extensions/data-workspace/src/services/workspaceService.ts

```diff
@@ -0,0 +1,65 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import * as dataworkspace from 'dataworkspace';
+import * as path from 'path';
+import { IWorkspaceService } from '../common/interfaces';
+import { ProjectProviderRegistry } from '../common/projectProviderRegistry';
+
+const WorkspaceConfigurationName = 'dataworkspace';
+const ProjectsConfigurationName = 'projects';
+
+export class WorkspaceService implements IWorkspaceService {
+	async getAllProjectTypes(): Promise<dataworkspace.IProjectType[]> {
+		await this.ensureProviderExtensionLoaded();
+		const projectTypes: dataworkspace.IProjectType[] = [];
+		ProjectProviderRegistry.providers.forEach(provider => {
+			projectTypes.push(...provider.supportedProjectTypes);
+		});
+		return projectTypes;
+	}
+
+	async getProjectsInWorkspace(): Promise<string[]> {
+		if (vscode.workspace.workspaceFile) {
+			const projects = <string[]>vscode.workspace.getConfiguration(WorkspaceConfigurationName).get(ProjectsConfigurationName);
+			return projects.map(project => path.isAbsolute(project) ? project : path.join(vscode.workspace.rootPath!, project));
+		}
+		return [];
+	}
+
+	async getProjectProvider(projectFilePath: string): Promise<dataworkspace.IProjectProvider | undefined> {
```

Comment:
If I'm reading the code correctly, this is doing multiple passes through the list of providers that both identify something via a supported project type.  I'm probably being too hopeful here, but is there any way to jump from the vscode extension to the provider so we don't have to scan twice?

Reply:
I've refactored the code a bit and introduced a getProviderByProjectType method on the registry. worst case (the provider extension is not loaded), we still have to call this method twice